### ISSUE TITLE
HAProxy 2.4.22 unable to load certificate chain from file

### ISSUE THREAD

**Issue by cydex2018** [haproxy/haproxy#2791]
### Detailed Description of the Problem

Hello everyone,

since today's update to Rocky Linux 9.5, haproxy can no longer be started.

Nov 24 14:53:36 <srvname> haproxy[2313]: [ALERT] (2313) : parsing [/etc/haproxy/haproxy.cfg:59] : 'bind <virtip>:443' : unable to load certificate chain from file '/etc/haproxy/certs/site.pem'.
Nov 24 14:53:36 <srvname> haproxy[2313]: [ALERT] (2313) : Error(s) found in configuration file : /etc/haproxy/haproxy.cfg
Nov 24 14:53:36 <srvname> haproxy[2313]: [ALERT] (2313) : Fatal errors found in configuration.
Nov 24 14:53:36 <srvname> systemd[1]: haproxy.service: Control process exited, code=exited, status=1/FAILURE

The file /etc/haproxy/certs/site.pem contains the certificate + Intermediate (in that order)
The key is located in the file /etc/haproxy/certs/site.pem.key

Before the update, the service started with this configuration without any problems

### Expected Behavior

HAProxy should start without any problems (has worked like this in the past)

### Steps to Reproduce the Behavior

Rocky Linux 9.5
Splitting the certificate + intermediate and the private key into different files
- /etc/haproxy/certs/site.pem (Certificate + Intermediate)
- /etc/haproxy/certs/site.pem.key or /etc/haproxy/certs/site.key (Private Key)
- Apply Config and Start HAProxy

### Do you have an idea how to solve the issue?

If the private key is inserted in the file /etc/haproxy/certs/site.pem, the service starts without issues.

### What is your configuration?

```haproxy
global
    log         127.0.0.1 local2 debug

    chroot      /var/lib/haproxy
    stats socket /var/run/haproxy_admin.sock mode 660 level admin
    pidfile     /var/run/haproxy.pid
    maxconn     4000
    user        haproxy
    group       haproxy
    daemon

 
    stats socket /var/lib/haproxy/stats


    ssl-default-bind-ciphers PROFILE=SYSTEM
    ssl-default-server-ciphers PROFILE=SYSTEM
    ssl-default-bind-options ssl-min-ver TLSv1.2

defaults
    mode                    http
    log                     global
    option                  httplog
    option                  dontlognull
    option http-keep-alive
    option forwardfor       except 127.0.0.0/8
    option                  redispatch
    retries                 3
    timeout http-request    10s
    timeout queue           1m
    timeout connect         10s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 3000

frontend stats
    mode http
    bind <ext-ip>:8404
    stats enable
    stats uri /stats
    stats refresh 10s

frontend <srvname>
    bind <virt-ip>:443 ssl crt /etc/haproxy/certs/site.pem
    default_backend             <backendname>

backend <backendname>
    balance leastconn
    mode http

    cookie SRV_USE insert indirect nocache dynamic
    dynamic-cookie-key <key>

    server <srvname> <fqdn>:443 ssl verify none check source <same_as_virtip>
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.22-f8e3218 2023/02/14 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.22.html
Running on: Linux 5.14.0-503.14.1.el9_5.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Nov 15 12:04:32 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL +EPOLL -EVPORTS +FUTEX +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OT -PCRE +PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PRIVATE_CACHE -PROCCTL +PROMEX -PTHREAD_PSHARED -QUIC +RT +SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 3.0.7 1 Nov 2022
Running on OpenSSL version : OpenSSL 3.2.2 4 Jun 2024 (VERSIONS DIFFER!)
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.4.1 20230605 (Red Hat 11.4.1-2)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
journalctl -xeu haproxy.service
Nov 24 14:53:36 <srvname> haproxy[2313]: [ALERT]    (2313) : parsing [/etc/haproxy/haproxy.cfg:59] : 'bind <virtip>:443' : unable to load certificate chain from file '/etc/haproxy/certs/site.pem'.
Nov 24 14:53:36 <srvname> haproxy[2313]: [ALERT]    (2313) : Error(s) found in configuration file : /etc/haproxy/haproxy.cfg
Nov 24 14:53:36 <srvname> haproxy[2313]: [ALERT]    (2313) : Fatal errors found in configuration.
Nov 24 14:53:36 <srvname> systemd[1]: haproxy.service: Control process exited, code=exited, status=1/FAILURE
```

### Additional Information

cat /etc/os-release
`NAME="Rocky Linux"
VERSION="9.5 (Blue Onyx)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="9.5"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Rocky Linux 9.5 (Blue Onyx)"
ANSI_COLOR="0;32"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:rocky:rocky:9::baseos"
HOME_URL="https://rockylinux.org/"
VENDOR_NAME="RESF"
VENDOR_URL="https://resf.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
SUPPORT_END="2032-05-31"
ROCKY_SUPPORT_PRODUCT="Rocky-Linux-9"
ROCKY_SUPPORT_PRODUCT_VERSION="9.5"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="9.5"
`

/etc/sysctl.d/00-vhl.conf (for virtual-ip keepalived)
`net.ipv4.ip_nonlocal_bind = 1`

**Comment by wlallemand** [comment]
> Built with OpenSSL version : OpenSSL 3.0.7 1 Nov 2022
> Running on OpenSSL version : OpenSSL 3.2.2 4 Jun 2024 (VERSIONS DIFFER!)

HAProxy does not run with the version it was built for, it could be a problem. Also check if your SECLEVEL changed if you use old algorithms in your certificates.

**Comment by cydex2018** [comment]
I think its related to openssl rebase 3.2.x in RHEL 9.5
Rocky 9.4 uses openssl-1:**3.0.7**-28.el9_4.x86_64
Rocky 9.5 openssl-1:**3.2.2**-6.el9_5.x86_64

**strace 9.5 (3.2.2-6.el9_5.x86_64)**
```
newfstatat(AT_FDCWD, "/etc/haproxy/certs/.", {st_mode=S_IFDIR|0755, st_size=42, ...}, 0) = 0
newfstatat(AT_FDCWD, "/etc/haproxy/certs/..", {st_mode=S_IFDIR|0755, st_size=113, ...}, 0) = 0
newfstatat(AT_FDCWD, "/etc/haproxy/certs/site.pem", {st_mode=S_IFREG|0644, st_size=2122, ...}, 0) = 0
openat(AT_FDCWD, "/etc/haproxy/certs/site.pem", O_RDONLY) = 4
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
fstat(4, {st_mode=S_IFREG|0644, st_size=2122, ...}) = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN CERTIFICATE-----\nMIIF"..., 4096) = 2122
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN CERTIFICATE-----\nMIIF"..., 4096) = 2122
read(4, "", 4096)                       = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN CERTIFICATE-----\nMIIF"..., 4096) = 2122
read(4, "", 4096)                       = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN CERTIFICATE-----\nMIIF"..., 4096) = 2122
read(4, "", 4096)                       = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN CERTIFICATE-----\nMIIF"..., 4096) = 2122
read(4, "", 4096)                       = 0
close(4)                                = 0
getpid()                                = 43852
brk(0x55f85afdf000)                     = 0x55f85afdf000
write(2, "[NOTICE]   (43852) : haproxy ver"..., 55[NOTICE]   (43852) : haproxy version is 2.4.22-f8e3218
) = 55
getpid()                                = 43852
write(2, "[NOTICE]   (43852) : path to exe"..., 57[NOTICE]   (43852) : path to executable is /sbin/haproxy
) = 57
getpid()                                = 43852
write(2, "[ALERT]    (43852) : parsing [/e"..., 163[ALERT]    (43852) : parsing [/etc/haproxy/haproxy.cfg:59] : 'bind <virtip>:443' : unable to load certificate chain from file '/etc/haproxy/certs/site.pem'.
) = 163
```
No private key will be loaded

**strace 9.4 (3.0.7-28.el9_4.x86_64)**
```
newfstatat(AT_FDCWD, "/etc/haproxy/certs/.", {st_mode=S_IFDIR|0755, st_size=42, ...}, 0) = 0
newfstatat(AT_FDCWD, "/etc/haproxy/certs/..", {st_mode=S_IFDIR|0755, st_size=113, ...}, 0) = 0
newfstatat(AT_FDCWD, "/etc/haproxy/certs/site.pem", {st_mode=S_IFREG|0644, st_size=2122, ...}, 0) = 0
openat(AT_FDCWD, "/etc/haproxy/certs/site.pem", O_RDONLY) = 4
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
fstat(4, {st_mode=S_IFREG|0644, st_size=2122, ...}) = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN CERTIFICATE-----\nMIIF"..., 4096) = 2122
lseek(4, 2122, SEEK_SET)                = 2122
read(4, "", 4096)                       = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN CERTIFICATE-----\nMIIF"..., 4096) = 2122
read(4, "", 4096)                       = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN CERTIFICATE-----\nMIIF"..., 4096) = 2122
read(4, "", 4096)                       = 0
close(4)                                = 0
newfstatat(AT_FDCWD, "/etc/haproxy/certs/site.pem.key", {st_mode=S_IFREG|0644, st_size=3244, ...}, 0) = 0
openat(AT_FDCWD, "/etc/haproxy/certs/site.pem.key", O_RDONLY) = 4
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
fstat(4, {st_mode=S_IFREG|0644, st_size=3244, ...}) = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "-----BEGIN RSA PRIVATE KEY-----\n"..., 4096) = 3244
close(4)                                = 0
```
Private key is loaded from /etc/haproxy/certs/site.pem.key

Cert is sha256WithRSAEncryption with 4096bit RSA Key from Let's Encrypt

**Comment by wlallemand** [comment]
Did you rebuild  haproxy for the 3.2 version ?
Also 2.4.22 is a bit old, you should update to 2.4.28, we won't debug an old version like this.

**Comment by cydex2018** [comment]
I have rebuild haproxy for 3.1-dev14 and now it is running with the same config
```
HAProxy version 3.1-dev14-c5d0342 2024/11/21 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 5.14.0-503.14.1.el9_5.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Nov 15 12:04:32 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CC      = cc
  CFLAGS  = -O2 -g -fwrapv
  OPTIONS = USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_PROMEX=1 USE_PCRE2=1
  DEBUG   =

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
Running on OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.5.0 20240719 (Red Hat 11.5.0-2)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=SPOP  side=BE     mux=SPOP  flags=HOL_RISK|NO_UPG
       spop : mode=SPOP  side=BE     mux=SPOP  flags=HOL_RISK|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```
Output:
```
[root@<srvname> haproxy-3.1-dev14]# ./haproxy -D -f /etc/haproxy/haproxy.cfg
[root@<srvname> haproxy-3.1-dev14]# 
```

3.0.6 is also running
```
HAProxy version 3.0.6-c2c0090 2024/11/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2029.
Known bugs: http://www.haproxy.org/bugs/bugs-3.0.6.html
Running on: Linux 5.14.0-503.14.1.el9_5.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Nov 15 12:04:32 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CC      = cc
  CFLAGS  = -O2 -g -fwrapv
  OPTIONS = USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1 USE_PCRE2=1
  DEBUG   =

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
Running on OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.5.0 20240719 (Red Hat 11.5.0-2)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```
Output:
```
[root@<srvname> haproxy-3.0.6]# ./haproxy -D -f /etc/haproxy/haproxy.cfg
[root@<srvname> haproxy-3.0.6]#
```

In 2.4.28 the problem still persists
```
HAProxy version 2.4.28-6b47cb7 2024/11/08 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.28.html
Running on: Linux 5.14.0-503.14.1.el9_5.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Nov 15 12:04:32 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL +EPOLL -EVPORTS +FUTEX +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OT -PCRE +PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PRIVATE_CACHE -PROCCTL +PROMEX -PTHREAD_PSHARED -QUIC +RT +SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
Running on OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.5.0 20240719 (Red Hat 11.5.0-2)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

Output:
```
[root@<srvname> haproxy-2.4.28]# ./haproxy -D -f /etc/haproxy/haproxy.cfg
[NOTICE]   (3677) : haproxy version is 2.4.28-6b47cb7
[NOTICE]   (3677) : path to executable is ./haproxy
[ALERT]    (3677) : parsing [/etc/haproxy/haproxy.cfg:59] : 'bind <ip>:443' : unable to load certificate chain from file '/etc/haproxy/certs/site.pem'.
[ALERT]    (3677) : Error(s) found in configuration file : /etc/haproxy/haproxy.cfg
[ALERT]    (3677) : Fatal errors found in configuration.
[root@<srvname> haproxy-2.4.28]#

```

**Comment by wlallemand** [comment]
Thanks for testing the different build, it looks like 2.4.28 is not behaving correctly with OpenSSL 3.2.2 indeed.
We have a reg-test which test this feature, I just run it and it's not working anymore with this version.
```
$ make reg-tests reg-tests/ssl/set_ssl_cert_noext.vtc                              

########################## Preparing to run tests ##########################
Testing with haproxy version: 2.4.28
Target : linux-glibc
Options : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL +EPOLL -EVPORTS +FUTEX +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY -LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OT -PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PRIVATE_CACHE -PROCCTL -PROMEX -PTHREAD_PSHARED -QUIC +RT +SLZ -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB
Services : none
########################## Gathering tests to run ##########################
  Add test: reg-tests/ssl/set_ssl_cert_noext.vtc
########################## Starting vtest ##########################
Testing with haproxy version: 2.4.28
#    top  TEST reg-tests/ssl/set_ssl_cert_noext.vtc TIMED OUT (kill -9)
#    top  TEST reg-tests/ssl/set_ssl_cert_noext.vtc FAILED (10.011) signal=9
1 tests failed, 0 tests skipped, 0 tests passed
########################## Gathering results ##########################
###### Test case: reg-tests/ssl/set_ssl_cert_noext.vtc ######
## test results in: "/tmp/haregtests-2024-11-25_21-44-13.ZmT8lo/vtc.459903.69d7c48f"

make: *** [Makefile:1115: reg-tests] Error 1
```

I don't think I will be able to fix this, because HAProxy 2.4 uses an API which is deprecated since OpenSSL 3.0, so there is a chance that this deprecated API is the cause of the problem, and broke in 3.2. HAProxy 2.6, which was ported to the new API, does not have this problem.

**Comment by wlallemand** [comment]
I checked again, and the behavior of OpenSSL changed during the 3.0 version cycle, I have the same problem with the 3.0.14 from ubuntu, the error code changed when trying to load a type which is not in the PEM file. This was fixed when porting to OpenSSL 3.0 API in HAProxy 2.6, but it was never fixed in 2.4. I  made a fix https://git.haproxy.org/?p=haproxy-2.4.git;a=commit;h=d03501c1bab66283f143ff8629db7d7f62d3f4ad . However HAProxy 2.4 releases are for critical fixes only, so don't expect a new release soon! Also OpenSSL >= 3.0 is not officially supported for this version. So I suggest that you upgrade to at least branch 2.6.

**Comment by sknolin** [comment]
We stumbled across this. We found (by guessing) that we could make this load the chain file if we we went back to the old "add the private key to the certificate pem file" instead of relying on the foo.name.key separate file. Leaving this here as it's a convenient workaround temporarily.

**Comment by NeuRA-Labs** [comment]
We have also noticed this since we did an update on two of our Rocky linux 9.5 servers. OpenSSL was updated via dnf automatic to OpenSSL 3.2.2 after which we started getting "unable to load certificate chain from file" errors. We employed the same fix as @sknolin going back to the old way of concat'ing the fullchain and privkey. We'd prefer to go back to having the certs separate.

Given Rocky linux is using upstream EL packages, is there any way we can let Red hat know to see if they can bump the version of HAProxy from 2.4 to 2.6 to fix this? 

I'm guessing our best bet for the moment is to install a community maintained version from https://packages.zenetys.com/latest/redhat/9/RPMS/x86_64/ yeah?
